### ISSUE TITLE
Variations of load, missingExecutables, and Documentation

### ISSUE THREAD

**Issue by ScottFreeCode** [luke-clifton/shh#80]
TL;DR:

1. Some suggested documentation of the relationship of the `load…` family of functions
2. A question/suggestion about the implementation of `load[Annotated]Env` (relative to `loadFromDirs`)
3. A suggestion to add the runtime check function (`missingExecutables`) to the functions that find exes at compile-time.

## Load Function Variations

I've put together a couple charts of the relationship of the `load…` functions!

### Family Tree

Which functions are called by which?

```text
exe
 |
 \__rawExe†
     |
     \_____loadExeAs
     |      |
     |      \________loadExe
     |      |
     |      \________loadAnnotated
     |                |
     |                \____________load
     |
     \_____loadAnnotatedFromDirs
     |      |
     |      \____________________loadFromDirs
     |                            |
     |                            \___________loadFromBins
     |
     \_____loadAnnotatedEnv
            |
            \_______________loadEnv
```

### Feature Chart

|Name |Template Haskell|One or Many|Compile-Time Check|Build Runtime Check Function|Search Dirs Find Exes (Compile-Time)|Name Transform Function|
|-----|----------------|-----------|------------------|----------------------------|------------------------------------|-----------------------|
|`exe`|      No        |One        |No                |No                          |No                                  |None (name parameter)  |
|`rawExe`|   Yes†      |One        |No                |No                          |No                                  |None (name parameter)  |
|`loadExeAs`|Yes       |One        |Yes               |No                          |No                                  |None (name parameter)  |
|`loadExe`|  Yes       |One        |Yes               |No                          |No                                  |None (name matches exe)|
|`load`|     Yes       |Many       |Yes               |Yes                         |No                                  |Default                |
|`loadAnnotated`|Yes   |Many       |Yes               |Yes                         |No                                  |User-supplied          |
|`loadFromDirs`| Yes   |Many       |No                |No                          |User-Supplied                       |Default                |
|`loadFromBins`| Yes   |Many       |No                |No                          |User-Supplied <$> (<> /bin)         |Default                |
|`loadAnnotatedFromDirs`|Yes|Many  |No                |No                          |User-Supplied                       |User-supplied          |
|`loadEnv`|         Yes|Many       |No                |No                          |PATH                                |Default                |
|`loadAnnotatedEnv`|Yes|Many       |No                |No                          |PATH                                |User-supplied          |

† In both charts, `rawExe` has been included for its Template Haskell even though it is _not exposed_ – `rawExe fName executable` is Template Haskell for `fName = exe executable`

## loadEnv vs loadFromDirs

If I'm not missing something in my feature chart, it almost looks to me as if `load[Annotated]Env` could be implemented in terms of `load[Annotated]FromDirs` and simply pass it the directories in the split `PATH`.

If there's another difference, it would be good to know. Otherwise, it might simplify things slightly (in terms of the implementation and tracing the relationships in the code) to implement the one in terms of the other.

That would make the family tree this instead:

```text
exe
 |
 \__rawExe†
     |
     \_____loadExeAs
     |      |
     |      \________loadExe
     |      |
     |      \________loadAnnotated
     |                |
     |                \____________load
     |
     \_____loadAnnotatedFromDirs
            |
            \____________________loadFromDirs
            |                     |
            |                     \___________loadFromBins
            |
            \____________________loadAnnotatedEnv
                                  |
                                  \_______________loadEnv
```

## Building the Runtime Check Function

It makes sense that the `load` functions _either_ compile-time check for the exe _or_ find exes in directories by searching at compile time.

It looks as though the ones that search, however, are missing building the runtime check function `missingExecutables`.

That would be a nice addition to align the runtime check feature with the loads multiple exes feature. (The compile-time and runtime check columns in the chart could be swapped to make it obvious.)

### Asides

- The Readme says, "Automatically defining a function for each executable on your `$PATH` using template Haskell, as well as a runtime check to ensure they all exist on startup." But as per the above, the Env functions don't generate the `missingExecutables` function. (I tested this out to be sure, neither FromDirs nor Env provide it whereas `load` does.)
- The function documentation mentions in `rawExe` that it does not perform the compile-time check for the executable, but does not mention which functions do nor does `loadExeAs` (which implements it) document that it does so (which also gets called by `loadExe`, `load` and `loadAnnotated`).
- It might be handy to have an example of using `missingExecutables`. The description and type signature are enough to figure out how to use it, but I thought I saw a simple, suggested usage pattern and now I can't seem to find it again.
- As a mere matter of source code organization, I noticed that `endBy` is among the `load` functions instead of among the `endBy…` functions.
